Update single-tree-render codemod: unmountComponentAtNode

simison opened the pull request:
_Related to PR #19494 "Single tree-rendering-issues"_

Removes instances of:
```js
ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
```
...but only if that `secondary` element selector present; we only care about un-mounting the sidebar. There's no need to sniff target element's name since there are no `unmountComponentAtNode`'s for `primary` element in the codebase, apart from [this workaround](https://github.com/Automattic/wp-calypso/blob/master/client/boot/project/wordpress-com.js#L275), which will be removed manually.

Codemod also removes un-used ReactDom imports (preserving comments above them).

## Testing
#### 1. Removing correct `unmountComponentAtNode`s
```bash
npm run codemod single-tree-rendering \
./client/post-editor/controller.js \
./client/my-sites/checkout/controller.jsx \
./client/jetpack-connect/controller.js
```
→ Removes `ReactDom.unmountComponentAtNode()`

#### 2. Shouldn't remove all `unmountComponentAtNode`s
```bash
npm run codemod single-tree-rendering ./client/lib/accept/index.js
```
→ Doesn't remove `ReactDom.unmountComponentAtNode()` because it doesn't have `secondary` element selector.

#### 3. Removes imports
```bash
npm run codemod single-tree-rendering ./client/my-sites/customize/controller.js
```
→ Stranded reactDom import is removed.


#### 4. Doesn't remove all imports
```bash
npm run codemod single-tree-rendering ./client/my-sites/plugins/plugin-list-header/index.jsx
```
→ Does not remove `import { findDOMNode } from 'react-dom';`

#### 5. Doesn't remove comments above removed imports
```bash
npm run codemod single-tree-rendering \
./client/auth/controller.js \
./client/me/happychat/index.jsx \
./client/my-sites/plans/current-plan/controller.jsx
```
→ Comments are preserved while imports are removed under them

ockham commented:
> Transforms instances of:
> ```js
> ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
> ```
> ...to this:
> ```js
> context.secondary = null;
> ```

I was actually [considering](https://github.com/Automattic/wp-calypso/pull/19567#issuecomment-343864805) ripping it out altogether, i.e. without replacement.

simison commented:
@ockham Ah, right on! Thanks for the heads up, I somehow missed that from your comment.

I updated the PR to remove them instead. This will also make it much simpler because we don't need to worry about `context` and `next()` anymore.

Do you think we could trust tests to tell us if there's a sidebar when there shouldn't?

We can now probably clean out these as well:
```js
context.store.dispatch(
  setSection(
    { name: 'jetpackConnect' },
    {
      hasSidebar: false,
    }
  )
);
```
...if they don't set anything else but hasSidebar. I might be missing something here tho, so I can't say for sure?

simison commented:
Updated this not to remove just any `react-dom` imports. E.g. [this should remain](https://github.com/Automattic/wp-calypso/blob/4d7857c8c68afffe7f9eeefb2c7e816f31060450/client/my-sites/plugins/plugin-list-header/index.jsx#L11):
```js
import { findDOMNode } from 'react-dom';
```

Updated also testing instructions.

simison commented:
I abstracted removing imports while preserving comments above them.

@ockham I believe this is now ready for review!

simison commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -78,6 +84,38 @@ export default function transformer( file, api ) {
+		const comment = _.get( node, 'comments[0]', false );
+
+		// Remove import (and any comments with it)
+		collection.remove();
```
Note that this wouldn't work: `node.remove()` so I'm removing the whole collection, although they will always contain only one `import` node.

simison commented:
Hmm, seems like I overlooked situations where there might be multiple comments above removed imports. Silly me. Fixed.

ockham commented:
> We can now probably clean out these as well:
> ```js
> context.store.dispatch(
>   setSection(
>     { name: 'jetpackConnect' },
>     {
>       hasSidebar: false,
>     }
>   )
> );
> ```
> ...if they don't set anything else but hasSidebar. I might be missing something about those `setStore()`s though, so I can't say for sure.

(You mean `setSection` :slightly_smiling_face: )

I wish we could, but that action is dispatched to set a `className` to enable some required styling. Here's some previous discussion on what we might try to actually get rid of it: https://github.com/Automattic/wp-calypso/pull/14812#issuecomment-306763017

ockham commented:
> I believe this is now ready for review! 👌 

BTW I think you should be able to change the GH Label from 'In Progress' to 'Needs Review' to signal that...

ockham commented:
> Do you think we could trust tests to tell us if there's a sidebar when there shouldn't?

We don't have tests in place to cover that, I'm afraid.

In fact, with this PR, we might be approaching a point where we need to run the updated codemod more often on the codebase and actually test whether Calypso still works -- in this specific case, whether switching back and forth between sections with and without sidebar works, or if there are weird rendering errors (e.g. sidebar persists in  a section that shouldn't have any).

Might be worth briefly looking into whether it'd be worth writing an [e2e test](https://github.com/Automattic/wp-e2e-tests) to take some of the manual testing burden from us.

simison commented:
Right, sorry for being unclear. :-)

>I wish we could, but that action is dispatched to set a className to enable some required styling. Here's some previous discussion on what we might try to actually get rid of it: #14812 (comment)

I meant that if we later on refactor not to check for `hasSidebar` (but `secondary` instead), is there any other reason to dispatch this action from middleware? Sometimes I've seen that action being dispatched with `name: null`.

Basically, would this make any sense/difference:
```js
context.store.dispatch(
  setSection(
    { name: 'jetpackConnect' },
    { }
  )
);
```

We should leave this outside the STR scope for now anyway but for me it's just interesting to understand. :-)

ockham commented:
> I meant that if we later on refactor not to check for `hasSidebar` (but `secondary` instead),

Ah right, we discussed that before. Sry, I forgot for a moment.

> I meant that if we later on refactor not to check for `hasSidebar` (but `secondary` instead), is there any other reason to dispatch this action from middleware? Sometimes I've seen that action being dispatched with `name: null`.

> Basically, would this make any sense/difference:
> ```js
> context.store.dispatch(
>   setSection(
>     { name: 'jetpackConnect' },
>     { }
>   )
> );
> ```

It only makes sense when `name` changes, i.e. when switching between sections (which is done automagically via `server/bundler/loader` IIRC). If the call was previously only there to set `hasSidebar` (which as you noticed often goes in hand with `name: null`), it should probably be removed.

> We should leave this outside the STR scope for now anyway but for me it's just interesting to understand. :-)

👍

simison commented:
>In fact, with this PR, we might be approaching a point where we need to run the updated codemod more often on the codebase and actually test whether Calypso still works -- in this specific case, whether switching back and forth between sections with and without sidebar works, or if there are weird rendering errors (e.g. sidebar persists in a section that shouldn't have any).
>
>Might be worth briefly looking into whether it'd be worth writing an e2e test to take some of the manual testing burden from us.

My thoughts precisely. :thumbsup:

Adding tests to check presence/absence of the sidebar would be helpful indeed.

I've been running/testing the codemod since yesterday against fresh master but kept hitting eslint errors. I'll push fresh run to #19494 soon.

BTW I suppose that weird timeout error from canary tests isn't anything to worry about?

simison commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -78,6 +84,39 @@ export default function transformer( file, api ) {
+		if ( comments.length ) {
+			const isRemovedExternal = isExternal( node );
+
+			// Find remaining external or internal dependencies and place comments above first one
```
We might still wipe out `/** @format */` comments if there aren't any imports left where we could place comments back.

I wouldn't worry about this until we notice it's actually happening.

ockham commented:
> I've been running/testing the codemod since yesterday against fresh master but kept hitting eslint errors. 

What kind of? If they're unrelated and/or irrelevant (and just happen to be triggered because they're on the lines we modify), we might just ignore them. It's clearly more important that this thing works than that we're eslint free (we shouldn't introduce avoidable new ones tho).

ockham commented:
Oh BTW, with the Node 8 upgrade in place, it'd be probably a good idea to rebase our open PRs...

ockham commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -256,6 +295,33 @@ export default function transformer( file, api ) {
+		},
+	} );
+
+	// Remove stranded `reactDom` imports
```
Nitpick: Should be either `ReactDom` or `react-dom` in the comment. (I kinda stumble over it when reading, might be my OCD 😄)

ockham commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -256,6 +295,33 @@ export default function transformer( file, api ) {
 		.replaceWith( ensureNextMiddleware )
 		.forEach( transformReactDomRender );
 
+	// Find if `reactDom` is used
```
Nitpick: `ReactDom`, see below.

ockham commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -78,6 +84,39 @@ export default function transformer( file, api ) {
+	 *
+	 * @param {object} collection Collection containing at least one node. Comments are preserved only from first node.
+	 */
+	function removeImport( collection ) {
```
I've always wondered why a handy tool like this isn't part of upstream -- `react-codemod`, specifically. Or is it?

simison replied:
Good question! I had a look at `react-codemod` now and didn't see anything like this, but it's a good idea to try more actively look for generic helpers like this.

I might eventually move generic helpers like this to a separate codemod-utils file.

ockham commented:
> ## Testing
> #### 1. Removing correct `unmountComponentAtNode`s
> ```bash
> npm run codemod single-tree-rendering \
> ./client/post-editor/controller.js \
> ./client/my-sites/checkout/controller.jsx \
> ./client/jetpack-connect/controller.js
> ```
> → Removes `ReactDom.unmountComponentAtNode()`

ockham commented:
> #### 2. Shouldn't remove all `unmountComponentAtNode`s
> ```bash
> npm run codemod single-tree-rendering ./client/lib/accept/index.js
> ```
> → Doesn't remove `ReactDom.unmountComponentAtNode()` because it doesn't have `secondary` element selector.

No longer seems to replace the `ReactDom.render()` call with  `context.primary =` tho...

ockham commented:
^ I might've found two rather relevant bugs, LMK if you can repro...

ockham commented:
I also wonder if we're hitting a degree of complexity where we want to cover our functionality with some tests to guard against regressions. Some sort of snapshot testing would be cool I guess, where we just provide snapshots that define what the resulting code should look like...

simison commented:
>What kind of? If they're unrelated and/or irrelevant (and just happen to be triggered because they're on the lines we modify)

I simply solved them with modifications in this PR rather than fixing them directly there. Stuff like `react-dom` import (😉 ) left behind when it's not used anymore. It wasn't really stopping me as I could've just not cherry-picked this PR in that run, but I felt like leaving this out from the run wasn't really productive neither.

I did the run and pushed it already today because I didn't want to wait for this PR any longer: https://github.com/Automattic/wp-calypso/pull/19494 — and will rebase+rerun again with Node 8 once this is merged (and re-run more frequently from now on).

simison commented:
>Hmm I just tried this but it doesn't seem to actually remove ReactDom.unmountComponentAtNode() from those files 🤔

Oh, that was weird, I had missing `[0]` from [rather unsexy `_.get()`](https://github.com/Automattic/wp-calypso/blob/a16804ab75b6ba6af40f411ed1b1d66eb8e12d22/bin/codemods/src/single-tree-rendering.js#L377) — I thought tested this PR but apparently I didn't (it has been pretty hectic past two days with kids). Anyways, fixed & double checked now. :thumbsup: My bad, sorry for the trouble!

> No longer seems to replace the ReactDom.render() call with context.primary = tho...

Ah, sorry again, this specific case in `client/lib/accept/index.js` is known to fail, should've warned you about that! Uncommon usages of `ReactDom.unmountComponentAtNode` tend to come with uncommon usages of `ReactDom.render`: it's not in a middleware and it doesn't have `document.getElementById('primary')`. There are few of those cases and some of them [we can probably let be like this](https://github.com/Automattic/wp-calypso/pull/19567#issuecomment-343190749). I'm keeping a list of problematic ones at #18231

simison commented:
This one is pretty good test file because it contains both `ReactDom.render()` as well `ReactDom.unmountComponentAtNode()`:

```bash
npm run codemod single-tree-rendering ./client/devdocs/controller.js
```

simison commented:
FYI, rebased against master (i.e. this branch is on Node.js 8 now)

ockham commented:
> Ah, sorry again, this specific case in `client/lib/accept/index.js` is known to fail, should've warned you about that! Uncommon usages of `ReactDom.unmountComponentAtNode` tend to come with uncommon usages of `ReactDom.render`: it's not in a middleware and it doesn't have `document.getElementById('primary')`. There are few of those cases and some of them [we can probably let be like this](https://github.com/Automattic/wp-calypso/pull/19567#issuecomment-343190749). I'm keeping a list of problematic ones at #19494

Right, I could've thought of that myself, given that `lib/accept` renders a dialog.

There's one more thing I've noted when running the codemod on that file -- it adds a `context` arg to the default export tho it doesn't really need one. Ideally we can get rid of that in a separate PR.

simison commented:
>great testing instructions!

Cheers. :-) My extra 5 mins spent on testing instructions tends to save tons of everyone else's time.

>I also wonder if we're hitting a degree of complexity where we want to cover our functionality with some tests to guard against regressions. Some sort of snapshot testing would be cool I guess, where we just provide snapshots that define what the resulting code should look like...

Yeah I was wondering that too. It feels weird to write code and not write any testing. :-) On the other hand frequent runs of the codemod tend to reveal regressions pretty fast, so I wouldn't be overly worried. We'll also have _very_ close look at codemod's results.

I've been eyeing [jscodeshift-ava-tester](https://github.com/jfmengels/jscodeshift-ava-tester)'s structure (could be any other tester than AVA of course). The problem is that testing codemods using strings comparisons outdates even faster than codemods they're testing...
